bouncer/shared/utils.ts

```diff
@@ -294,84 +294,107 @@ export async function observeSwapEvents(
   { sourceAsset, destAsset, depositAddress, channelId }: SwapParams,
   api: ApiPromise,
   tag?: string,
-  swapType?: SwapType,
   finalized = false,
 ): Promise<BroadcastId | undefined> {
   let eventFound = false;
   const subscribeMethod = finalized
     ? api.rpc.chain.subscribeFinalizedHeads
     : api.rpc.chain.subscribeNewHeads;
 
+  const swapRequestedEvent = 'SwapRequested';
   const swapScheduledEvent = 'SwapScheduled';
   const swapExecutedEvent = 'SwapExecuted';
   const swapEgressScheduled = 'SwapEgressScheduled';
   const batchBroadcastRequested = 'BatchBroadcastRequested';
-  let expectedMethod = swapScheduledEvent;
+  let expectedMethod = swapRequestedEvent;
 
-  let swapId = 0;
+  let swapId: number | undefined;
+  let swapRequestId: number | undefined;
   let egressId: EgressId;
   let broadcastId;
 
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   const unsubscribe: any = await subscribeMethod(async (header) => {
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
     const events: any[] = await api.query.system.events.at(header.hash);
-    events.forEach((record) => {
-      const { event } = record;
-      if (!eventFound && event.method.includes(expectedMethod)) {
+
+    // Some events don't appear in the order one might expect (e.g. SwapRequested
+    // comes after SwapScheduled), so as a workaround, we add an other loop to be
+    // able to re-scan events that we already visited:
+    let keepSearching = true;
+
+    while (keepSearching) {
+      keepSearching = false;
+
+      for (const record of events) {
+        const { event } = record;
+        if (eventFound || !event.method.includes(expectedMethod)) {
+          // eslint-disable-next-line no-continue
+          continue;
+        }
+
         const expectedEvent = {
           data: event.toHuman().data,
         };
 
+        const data = expectedEvent.data;
+
         switch (expectedMethod) {
+          case swapRequestedEvent: {
+            const channel = data.origin.DepositChannel;
+
+            if (
+              channel &&
+              Number(channel.channelId) === channelId &&
+              Object.values(channel.depositAddress)[0] === depositAddress &&
+              sourceAsset === (data.inputAsset as Asset) &&
+              destAsset === (data.outputAsset as Asset)
+            ) {
+              swapRequestId = data.swapRequestId;
+              expectedMethod = swapScheduledEvent;
+
+              // This ensures that we loop through all events again to find SwapScheduled:
+              keepSearching = true;
```

Comment:
What do you mean by "recursion" exactly in this case? What would the signature of the function look exactly? One of my goals for this test was to fix it with minimal changes. IMO the concept of repeating the search in case we found something is not too difficult to understand.

Also the reason I don't want to spend time improving this code is that `observeSwapEvents` (along with `performAndTrackSwap` and others) only exist to serve one test and should probably be removed in favour of more general approach for tracking events.